### ISSUE TITLE
Better integration for VMs and Gateways

### ISSUE THREAD

**Issue by isaacabraham** [CompositionalIT/farmer#280]
I've cleaned up some of this so you can more easily connect a VM to a Gateway:

```fsharp
let developmentEnvironment = vm {
    name "dev-jan"
    username "jan"
    vm_size Vm.Standard_F8s_v2
    operating_system Vm.WindowsServer_2019Datacenter
    os_disk 128 Vm.StandardSSD_LRS
    add_ssd_disk 512
}

let gw = gateway {
    name (developmentEnvironment.Name.Map(sprintf "%s-gateway"))
    gateway_ip_config DynamicPrivateIp developmentEnvironment.IpName
    vnet developmentEnvironment.VNetName.ResourceName
    vpn_gateway_sku VpnGatewaySku.Basic
}
```

However we could go further:

```fsharp
let gw = gateway {
    link_to_vm developmentEnvironment // does the same as the name, gateway_ip_config and vnet keywords above
    vpn_gateway_sku VpnGatewaySku.Basic
}
```

Or even

```fsharp
let developmentEnvironment = vm {
    name "dev-jan"
    username "jan"
    vm_size Vm.Standard_F8s_v2
    operating_system Vm.WindowsServer_2019Datacenter
    os_disk 128 Vm.StandardSSD_LRS
    add_ssd_disk 512
    use_gateway VpnGatewaySku.Basic // don't need a separate gateway builder, just turn it on with sane defaults here...
}
```

Question is, is this the "right" thing to do? Are the defaults above (especially the choice of DynamicPrivateIp) sane? I don't mind if it doesn't suit everyone - I'm thinking of 80% of the people that want a gateway.

@Jandev As a user of Farmer and someone who just did something like this, which of the three approaches above would you find most logical - we could even include all three to fit all use cases.

**Comment by isaacabraham** [comment]
Also cc: @ninjarobot incase you have some thoughts?

**Comment by Jandev** [comment]
If I was to choose between these three options, I'd go for the third one.

The current `vm`-builder adds a public IP to the virtual machine. While this is the default behavior, it's not recommended to expose machines to the internet without any security in place. I get why it's developed like this now, but seeing the gateway is supported now, I think it's better to integrate this stuff with the builder and have a better solution out of the box.  

That said, I think people in actual projects will probably not use the builders a lot as they want to be 'in control' to what gets created.  
Having the possibility to do the first option will probably be something I'd use in an actual project. That way I know what gets created and can finetune it when necessary.

In short: 
* Option 3 is great for proof of concepts and get you started quickly
* Option 1 is great for actual projects when you need to understand what happens and need the versatility to tune your resources

**Comment by isaacabraham** [comment]
@Jandev great, thanks for the advice. I think we can support all of those options.

Having the gateway "on" as the default comes with one cost - the deployment time goes through the roof, so I would be reticent to turn that on by default, but having a simple turn-key solution for "I want a VM that is only accessible via a private vnet + gateway" seems like a good compromise.

In my current branch, this *seems* to be the sort of "manual" experience we have:

```fsharp
let devVn = vnet {
    name "dev-isaac-vnet"
    add_address_spaces [ "10.0.0.0/16" ]
    add_subnets [
        subnet { name "GatewaySubnet"; prefix "10.0.0.0/24" }
        subnet { name "Frontend"; prefix "10.0.2.0/24" }
    ]
}

let developmentEnvironment = vm {
    name "dev-isaac"
    username "isaac"
    link_to_vnet devVn
    subnet_name "Frontend"
}

let gw = gateway {
    name "dev-isaac-gateway"
    vnet devVn
    vpn_gateway_sku VpnGatewaySku.Basic
}
```

I must admit, even for this approach, I'm not especially fond of the code above. I would much rather have a flag on the VNet that simply said "make gateway" rather than having to create a separate gateway and then link it to the vnet. @ninjarobot what do you think? Or at least the ability to easily create the "magic" GatewaySubnet" in a sane, safe manner. There's still a lot that people can easily screw up here - I spent a few hours yesterday fumbling around before I gave up.

Although it's not really demonstrated much, you can directly construct ARM resources using the code inside the `Farmer.Arm` namespace, which contains records that map more-or-less 1:1 with ARM resources, so there's always that escape hatch if needed.
